components/shared/webrender/viewport_description.rs

```diff
@@ -0,0 +1,183 @@
+    /// initial-scale
+    pub initial_scale: Scale<f32>,
+    /// minimum-scale
+    pub minimum_scale: Scale<f32>,
+    /// maximum-scale
+    pub maximum_scale: Scale<f32>,
+    /// user-scalable
+    pub user_scalable: UserZoom,
+}
+
+impl Default for ViewportDescription {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ViewportDescription {
+    /// Creates a Viewport
+    pub fn new() -> Self {
+        ViewportDescription {
+            width: ViewportLength::Auto,
+            height: ViewportLength::Auto,
+            initial_scale: Scale::new(DEFAULT_ZOOM),
+            minimum_scale: Scale::new(MIN_ZOOM),
+            maximum_scale: Scale::new(MAX_ZOOM),
+            user_scalable: UserZoom::Zoom,
+        }
+    }
+
+    /// Process key value pair in Viewport
+    pub fn process_viewport_key_value_pair(pairs: HashMap<String, String>) -> ViewportDescription {
+        let mut description = ViewportDescription::new();
+        for (key, value) in &pairs {
+            match key.as_str() {
+                "width" | "height" => {
+                    let length = Self::parse_viewport_value_as_length(value);
+                    if length != ViewportLength::Auto {
+                        if key == "width" {
+                            description.width = length;
+                        } else {
+                            description.height = length;
+                        }
+                    }
+                },
+                "initial-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.initial_scale = zoom;
+                    }
+                },
+                "minimum-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.minimum_scale = zoom;
+                    }
+                },
+                "maximum-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.maximum_scale = zoom;
+                    }
+                },
+                "user-scalable" => {
+                    if let Some(user_zoom) = Self::parse_viewport_value_as_user_zoom(value) {
+                        description.user_scalable = user_zoom;
+                    }
+                },
+                _ => (),
+            }
+        }
+        description
+    }
+
+    /// Parses a viewport length value.
+    fn parse_viewport_value_as_length(value: &str) -> ViewportLength {
+        match value.to_lowercase().as_str() {
+            "device-width" => ViewportLength::DeviceWidth,
+            "device-height" => ViewportLength::DeviceHeight,
+            _ => value
+                .parse::<f32>()
+                .ok()
+                .filter(|&n| n >= 0.0)
+                .map(ViewportLength::Fixed)
+                .unwrap_or(ViewportLength::Auto),
+        }
+    }
```

Comment:
I think this could be an `Into<ViewportLength>` implementation https://doc.rust-lang.org/rust-by-example/conversion/from_into.html or `FromStr`.